lib/services/role.go

```diff
@@ -1887,69 +1616,155 @@ func (m *DatabaseNameMatcher) String() string {
 	return fmt.Sprintf("DatabaseNameMatcher(Name=%v)", m.Name)
 }
 
-// CheckAccessToDatabase checks if this role set has access to a particular database.
+// LoginMatcher is a RoleMatcher that checks whether the role's logins
+// match the specified condition.
+type LoginMatcher struct {
+	login *string
+}
+
+func NewLoginMatcher(login string) RoleMatcher {
+	return &LoginMatcher{login: &login}
+}
+
+func (l *LoginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
+	logins := role.GetLogins(typ)
+	for _, login := range logins {
+		if *l.login == login {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// AccessCheckable is the subset of types.Resource required for the RBAC checks.
+type AccessCheckable interface {
+	GetKind() string
+	GetName() string
+	GetMetadata() types.Metadata
+	GetAllLabels() map[string]string
+}
+
+// CheckAccess checks if this role set has access to a particular resource.
+// It attempts to match labels if roleLabelsFn is not nil.
 //
-// The checker always checks the server namespace, other matchers are supplied
-// by the caller.
-func (set RoleSet) CheckAccessToDatabase(server types.Database, mfa AccessMFAParams, matchers ...RoleMatcher) error {
+// Note: logging in this function only happens in debug mode. This is because
+// adding logging to this function (which is called on every resource returned
+// by the backend) can slow down this function by 50x for large clusters!
```

Comment:
nit: move this comment inside the function, before you create the `debugf` wrapper, I missed it on first pass.
Also, explain why your wrapper is much faster than `log.Debugf`. Also also, consider fixing `log.Debugf` upstream to do what your helper does.

Reply:
The helper is just to put the if statement in one place rather check the log level before every call to `log.Debugf`.

The comment and the associated code is not new, @russjones wrote it years ago. I'm not sure whether its necessary or if log.Debugf` is automatically a no-op when the level is not debug, but the comment sounds like this logic was created due to major performance issues, so I wanted to preserve the existing behavior.

Reply:
Moved comment inside, and fixed another occurrence of the same pattern in this file.